connectivity/tests/egressgateway.go

```diff
@@ -39,151 +155,249 @@ func (s *egressGateway) Name() string {
+	i = 0
+	for _, client := range ct.ClientPods() {
+		client := client
+
 		for _, externalEcho := range ct.ExternalEchoPods() {
-			t.NewAction(s, fmt.Sprintf("curl-%d", i), &client, externalEcho, check.IPFamilyV4).Run(func(a *check.Action) {
+			t.NewAction(s, fmt.Sprintf("curl-external-echo-pod-%d", i), &client, externalEcho, check.IPFamilyV4).Run(func(a *check.Action) {
 				a.ExecInPod(ctx, ct.CurlClientIPCommand(externalEcho, check.IPFamilyV4))
 				clientIP := extractClientIPFromResponse(a.CmdOutput())
 
-				if !clientIP.Equal(egressIP) {
+				if !clientIP.Equal(egressGatewayNodeInternalIP) {
 					t.Fatal("Request reached external echo service with wrong source IP")
 				}
 			})
 			i++
 		}
 	}
-}
 
-// getGatewayNodeInternalIP returns the k8s internal IP of the node acting as
-// gateway for this test
-func (s *egressGateway) getGatewayNodeInternalIP(ct *check.ConnectivityTest) net.IP {
-	gatewayNode, ok := ct.Nodes()[s.egressGatewayNode]
-	if !ok {
-		return nil
-	}
+	// When connecting from outside the cluster to a nodeport service whose pods are selected by an egress policy,
+	// the reply traffic should not be SNATed with the egress IP
+	i = 0
+	for _, client := range ct.ExternalEchoPods() {
+		client := client
 
-	for _, addr := range gatewayNode.Status.Addresses {
-		if addr.Type != "InternalIP" {
-			continue
+		for _, echo := range ct.EchoServices() {
+			// convert the service to a ServiceExternalIP as we want to access it through its external IP
+			echo := echo.ToExternalIPService()
+
+			t.NewAction(s, fmt.Sprintf("curl-echo-service-%d", i), &client, echo, check.IPFamilyV4).Run(func(a *check.Action) {
+				a.ExecInPod(ctx, ct.CurlClientIPCommand(echo, check.IPFamilyV4))
+			})
+			i++
 		}
+	}
 
-		ip := net.ParseIP(addr.Address)
-		if ip == nil || ip.To4() == nil {
-			continue
+	if status, ok := ct.Feature(check.FeatureTunnel); ok && !status.Enabled {
+		// When connecting from outside the cluster directly to a pod which is selected by an egress policy, the
+		// reply traffic should not be SNATed with the egress IP (only connections originating from these pods
+		// should go through egress gateway).
+		//
+		// This test is executed only when Cilium is running in direct routing mode, since we can simply add a
+		// route on the node outside the cluster to direct pod's traffic to the node where the pod is running
+		// (while in tunneling mode we would need the external node to send the traffic over the tunnel)
+		for _, echoPod := range ct.EchoPods() {
+			targetPodHostIP := echoPod.Pod.Status.HostIP
+			targetPodIP := echoPod.Pod.Status.PodIP
+
+			for _, externalNode := range ct.NodesWithoutCilium() {
+				for node, hostNetNSPod := range ct.HostNetNSPodsByNode() {
+					if node != externalNode {
+						continue
+					}
+
+					cmd := []string{"ip", "route", "add", targetPodIP, "via", targetPodHostIP}
```

Comment:
Could we use `WithIPRoutesFromOutsideToPodCIDRs()` instead? I know it's not fine-grained, but for your test case it should do the job.

Reply:
Nice, that works as well :rocket: